data/collector_types.go

```diff
@@ -0,0 +1,394 @@
+type CaptureResult struct {
+	// Type represents the type of result (binary or tabular)
+	Type CaptureType
+	// Timestamps contain the time the data was requested and received
+	Timestamps
+	// TabularData contains the tabular data payload when Type == CaptureResultTypeTabular
+	TabularData TabularData
+	// Binaries contains binary data responses when Type == CaptureResultTypeBinary
+	Binaries []Binary
+}
+
+// BEGIN CONSTRUCTORS
+
+// NewBinaryCaptureResult returns a binary capture result.
+func NewBinaryCaptureResult(ts Timestamps, binaries []Binary) CaptureResult {
+	return CaptureResult{
+		Timestamps: ts,
+		Type:       CaptureTypeBinary,
+		Binaries:   binaries,
+	}
+}
+
+// NewTabularCaptureResultReadings returns a tabular readings result.
+func NewTabularCaptureResultReadings(reqT time.Time, readings map[string]interface{}) (CaptureResult, error) {
+	var res CaptureResult
+	values, err := rprotoutils.ReadingGoToProto(readings)
+	if err != nil {
+		return res, err
+	}
+
+	return CaptureResult{
+		Timestamps: Timestamps{
+			TimeRequested: reqT,
+			TimeReceived:  time.Now(),
+		},
+		Type: CaptureTypeTabular,
+		TabularData: TabularData{
+			Payload: &structpb.Struct{
+				Fields: map[string]*structpb.Value{
+					"readings": structpb.NewStructValue(&structpb.Struct{Fields: values}),
+				},
+			},
+		},
+	}, nil
+}
+
+// NewTabularCaptureResult returns a tabular result.
+func NewTabularCaptureResult(reqT time.Time, i interface{}) (CaptureResult, error) {
+	var res CaptureResult
+	readings, err := protoutils.StructToStructPbIgnoreOmitEmpty(i)
+	if err != nil {
+		return res, err
+	}
+
+	return CaptureResult{
+		Timestamps: Timestamps{
+			TimeRequested: reqT,
+			TimeReceived:  time.Now(),
+		},
+		Type: CaptureTypeTabular,
+		TabularData: TabularData{
+			Payload: readings,
+		},
+	}, nil
+}
+
+// END CONSTRUCTORS
+
+// ToProto converts a CaptureResult into a []*datasyncPB.SensorData{}.
+func (cr *CaptureResult) ToProto() []*datasyncPB.SensorData {
+	ts := cr.Timestamps
+	if td := cr.TabularData.Payload; td != nil {
+		return []*datasyncPB.SensorData{{
+			Metadata: &datasyncPB.SensorMetadata{
+				TimeRequested: timestamppb.New(ts.TimeRequested.UTC()),
+				TimeReceived:  timestamppb.New(ts.TimeReceived.UTC()),
+			},
+			Data: &datasyncPB.SensorData_Struct{
+				Struct: td,
+			},
+		}}
+	}
+
+	var sd []*datasyncPB.SensorData
+	for _, b := range cr.Binaries {
+		sd = append(sd, &datasyncPB.SensorData{
+			Metadata: &datasyncPB.SensorMetadata{
+				TimeRequested: timestamppb.New(ts.TimeRequested.UTC()),
+				TimeReceived:  timestamppb.New(ts.TimeReceived.UTC()),
+				MimeType:      b.MimeType.ToProto(),
+				Annotations:   b.Annotations.ToProto(),
+			},
+			Data: &datasyncPB.SensorData_Binary{
+				Binary: b.Payload,
+			},
+		})
+	}
+	return sd
+}
+
+// Validate returns an error if the *CaptureResult is invalid.
+func (cr *CaptureResult) Validate() error {
+	var ts Timestamps
+	if cr.Timestamps.TimeRequested == ts.TimeRequested {
```

Comment:
I found these comparisons (to the unset `ts.Time*`) unintuitive; perhaps compare to a `var emptyTime time.Time` instead for readability?

Reply:
how is that?